pyesbulk/__init__.py

```diff
@@ -183,25 +304,79 @@ def actions_tracking_closure(cl_actions):
                     # ... otherwise consider it successful.
                     successes += 1
             elif status == 400:
-                doc = {
+                try:
+                    exc_payload = resp["exception"]
+                except KeyError:
+                    pass
+                else:
+                    # We have an exception object in the response object
+                    # which is not always JSON serializable, so we use
+                    # `repr` to turn that exception into a serializable
+                    # string while maintaining as much information about
+                    # the exception as possible.
+                    resp["exception"] = repr(exc_payload)
+                jsonstr = json.dumps(
+                    {
                         "action": action,
                         "ok": ok,
                         "resp": resp,
                         "retry_count": retry_count,
-                        "timestamp": tstos(time.time())
-                        }
-                jsonstr = json.dumps(doc, indent=4, sort_keys=True)
+                        "timestamp": _tstos(),
+                    },
+                    indent=4,
+                    sort_keys=True,
+                )
                 print(jsonstr, file=errorsfp)
                 errorsfp.flush()
                 failures += 1
             else:
-                # Retry all other errors
-                print(resp)
-                actions_retry_deque.append((retry_count + 1, action))
+                try:
+                    exc_payload = resp["exception"]
+                except KeyError:
+                    pass
+                else:
+                    resp["exception"] = repr(exc_payload)
+                try:
+                    error = resp["error"]
+                except KeyError:
+                    error = ""
```

Comment:
Here I think the subtle difference is that `error` is used in the `if` condition, where the `resp['exception']` value is just a payload fixup.

So we care that `error` is a valid string so that `error.startswith()` functions properly, but we don't care if `resp['exception']` exists or not, it is just dumped into a JSON string properly with the fixup we applied if it does exist.